### PR TITLE
[WIP] Lower foreach over List<T> to for-loop

### PR DESCRIPTION
**Pull request by bbartels**
# General

Lowering foreach iteration over List<T> to for loop, when applicable.

It came to me as quite the surprise when I found out that roslyn doesn't actually lower foreach loops to for
when iterating through a List<T>.  
Upon looking at the List<T> implementation I found out, that the reason why this isn't as trivial as for T[] is because List<T> keeps track of an internal version, which changes each time the list is modified.
As that internal version is not exposed publically this optimization would require a change in the List<T> API to allow the said version to be accessed.

By moving all the List<T>.Enumerator logic onto the callsite itself it would drastically increase performance in certain scenarios.
I would still need some further confirmation that this is actually a side-effect less optimization, I myself couldn't think of 
any possible things that could go wrong, but I may be completely missing something.

Currently using the Capacity property as a placeholder until the internal version of List<T> is exposed.

# Demonstration

## Initial C# code

```csharp
foreach (var x in some_list)
{
    // Do something with x
}
```

## Previous Roslyn translation 

```csharp
var enumerator = some_list.GetEnumerator();
try
{
    while (enumerator.MoveNext())
    {
        // Do something with enumerator.Current
    }
}
finally
{
    enumerator.Dispose();
}
```

## New roslyn translation

```csharp
int version = some_list.Version; // Version property does not currently exist
int i = 0;
while (i < some_list.Count)
{
    // Do something with some_list[i]
    if (version != some_list.Version)
    {
        throw new InvalidOperationException(); // Matching Exception message still needs to be passed to .ctor
    }
    i++;
}
```

# Benefits

Pro:
- Massive performance benefits when body of loop performs little work
    - In case of summing the length of all strings in List\<string\> up to 2.5x better performance
    - However, as the body of the loop does more work the benefits start to disintegrate
- ~~No allocation of Enumerator object~~ (Actually it shouldn't allocate as it is a struct Enumerator)

Con:
- Increased IL size (6 bytes per loop)
- Necessity of modifying the List\<T\> API in order to expose internal version 

## Benchmarks

This benchmark just sums up a list of integers for the first three results and the lengths of a list of strings for the last three.

# Considerations

There are a couple more things that have to be considered:

- Would this cause any side-effects?
- Would it cause issues when allowing lowering of types that derive from List\<T\>?
- Is it a good idea to expose the internal version field of List\<T\>?
    - If not, maybe a hashed version?
- Could/Should this be implemented in RyuJit to do runtime checking for List\<T\>, as e.g. IList\<T\> would not be lowered?

# Todo

Things that are yet to be completed, assuming aforementioned considerations are not going to be an issue:
- Add Tests
- Investigate/Implement possibility of allowing lowering of types that derive from List\<T\>
- Add API to expose internal version of List\<T\>
- Modify invoking InvalidOperationException constructor to take correct string argument
    - I am actually not quite sure how to do this as the no parameter constructor is already added as a WellKnownMember,
        so I can't add another one in WellKnownMembers.cs that takes one parameter, any help is appreciated!
- Modify PR to change WellKnownMember from System_Collections_Generic_List_T__get_Capacity to System_Collections_Generic_List_T__get_Version,
    once the Version API has been added

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
> I would still need some further confirmation that this is actually a side-effect less optimization, I myself couldn't think of any possible things that could go wrong, but I may be completely missing something.

As the implementation of `List<T>` could do anything, this is not a safe change to make.  It also violates teh language specification.  For this to be done it would have to be approved that hte language would have special knowledge of specific types like this, and the BCL team would have to approve providing semantics that would ensure all `List<T>` impls would have to be compliant.

**Comment by CyrusNajmabadi** [comment]
> It came to me as quite the surprise when I found out that roslyn doesn't actually lower foreach loops to for when iterating through a List.

`List<T>` is not a special type.  It's just a bog standard API  (like hundreds of other collections out there).  

--

For your benchmarks, can you include a gist with your actual test code in it?

**Comment by bbartels** [comment]
>  As the implementation of List<T> could do anything, this is not a safe change to make

I had considered that, however I came to the conclusion that because none of the Methods/Properties involved are marked as virtual, the enumeration behaviour of any List\<T\> implementation should always be the same, no? 

> It also violates teh language specification.

In what regard does it violate the specification? If List\<T\> is not a special API I don't see how it would expjicitly be part of the spec. Afaik the Span\<T\> lowering did not need such modifications to the spec.

**Comment by bbartels** [comment]
> For your benchmarks, can you include a gist with your actual test code in it?

See here, uses BenchmarkDotNet
https://pastebin.com/raw/Cfaq1cs6

**Comment by jcouv** [comment]
> Afaik the Span<T> lowering did not need such modifications to the spec.

LDM did make an explicit and narrow language design decision to recognize `Span`/`ReadOnlySpan` as a well-known type with regards to `foreach` ([here](https://github.com/dotnet/csharplang/blob/98043cdc889303d956d540d7ab3bc4f5044a9d3b/meetings/2017/LDM-2017-12-04.md#foreach-for-span-and-readonlyspan)).
That said, we don't have an updated consolidated spec which shows that, and looking at the existing spec I don't see previous optimizations (for example on array types) recorded there (but I expected that it would and should).

Those meeting notes also point out that we discussed a generalized feature allowing some types to be marked as "loopable", such that `foreach` would be lowered to a `for` for such types. But that feature was punted for now.

In short, the compiler team would not accept this PR (specific for `List` type) without an LDM decision. There are three options discussed so far: (1) stay as-is, (2) optimize `List` specifically, or (3) recognize "loopable" types. The last LDM decision was (1) following the `Span` optimization, but this PR is moving forward with (2).

The question is whether an LDM member feels strongly about (2) or (3) to champion either option. This is typically best handled on csharplang rather than with a PR. 
Also note that given our current focus on C# 8 and, particularly, nullability analysis at the moment, I don't expect we would discuss feature suggestions in LDM for another month or two.

Tagging @agocke @AlekseyTs @gafter in case anything to add.

**Comment by CyrusNajmabadi** [comment]
> I had considered that, however I came to the conclusion that because none of the Methods/Properties involved are marked as virtual, the enumeration behavior of any List<T> implementation should always be the same, no?

I could have a `List<T>` implementation that i personally write that does whatever I want.  This would change behavior.

> In what regard does it violate the specification?

The spec says how 'foreach' works (through calls to GetEnumerator and so forth), this would now be in violation of htat.

>  If List<T> is not a special API I don't see how it would explicitly be part of the spec. 

That's the problem.  If you want to say that a `foreach` over a `List<T>` needs to operate in a certain fashion, then you need to make it part of the spec.

**Comment by bbartels** [comment]
> > I had considered that, however I came to the conclusion that because none of the Methods/Properties involved are marked as virtual, the enumeration behavior of any List implementation should always be the same, no?
> 
> I could have a `List<T>` implementation that i personally write that does whatever I want. This would change behavior.

Ah, yeah you are right, I hadn't considered the fact that because List<T> implements the Enumerator<T> interface a derived class could just re-implement the GetEnumerator() call, consequently making this PR pretty useless. Should I just close this PR then?

**Comment by CyrusNajmabadi** [comment]
> Ah, yeah you are right, I hadn't considered the fact that because List implements the Enumerator interface a derived class could just re-implement the GetEnumerator() call, consequently making this PR pretty useless.

It's not even about having a *derived* class.  The literal `List<T>` impl could just be different and have different semantics depending on who wrote it.  I could provide my own `System.Collections.Generic.List<T>` that operates differently.  Unity could provide one that operates differently.  etc. etc. etc.

**Comment by bbartels** [comment]
> > Ah, yeah you are right, I hadn't considered the fact that because List implements the Enumerator interface a derived class could just re-implement the GetEnumerator() call, consequently making this PR pretty useless.
> 
> It's not even about having a _derived_ class. The literal `List<T>` impl could just be different and have different semantics depending on who wrote it. I could provide my own `System.Collections.Generic.List<T>` that operates differently. Unity could provide one that operates differently. etc. etc. etc.

That is a very fair point, now I understand what you mean! Sorry for my confusion, this is the first time I've done work in this sort of area, so everything is a little new to me :)

**Comment by CyrusNajmabadi** [comment]
You have nothing to be sorry about at all :)  It's a large and complex space, and it's great you're interested in it and want to improve things!

**Comment by RikkiGibson** [comment]
Seems related to dotnet/csharplang#2586

**Comment by gafter** [comment]
I believe the best approach to achieve this optimization is not to make the language recognize `List<T>` as special in any way, but rather to make `List<T>` provide its own custom iterator implementation using a value type.  And, look at that... it already does!

References:
- [A puzzle based on this fact](https://codeblog.jonskeet.uk/2010/07/27/iterate-damn-you/)
- [Demonstration of the code generated](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxNMAfAAgJgIwCwAUFgAwAEWeALANzHFYDMlO5AwuQN7Hl+Uss1cgFkIASwB2ACgCU3XvyUA3CAnIAbcbHIBecpLgB3cgBltMADxSYAPm7k8ScjmcsAvvSJKlAMwD2CHAQAMYAFuTSquoA1uRSmhayij4K3qk+VACc0jGyXhnk7ilFxO5AA). The inner loop is from `L0085` to `L0097`

Once your benchmark properly handles the "internal version", I suspect you are likely to see little improvement. But I could be wrong!

**Comment by agocke** [comment]
I'm going to close this out just because we don't currently have at least a preliminary language design decision that says we're interested in making this change. We can re-open if LDM takes the proposal.
